In [1]:
%load_ext autoreload
%autoreload 2

# Week 6 - Systematically Improving Your Rag Application

> **Prerequisites**: Make sure that you've completed the previous notebooks `1. Evaluate Tools` and `2. Generate Dataset` before continuing with this notebook. We'll be using the results from the previous notebook to evaluate the effectiveness of our new techniques.

In this notebook, we'll explore how to improve our model's ability to select the right tools using system prompts and few-shot examples.

## Why this matters

When deploying RAG systems that coordinate multiple tools, getting tool selection wrong wastes resources and degrades user experience. Simple techniques like system prompts and few-shot examples can significantly boost tool selection accuracy without complex infrastructure changes.

Just as Week 1 showed how synthetic data could improve retrieval, and Week 4 demonstrated how topic modeling helps understand query patterns, strategic prompting can help models better understand when and how to use different tools. By systematically testing these improvements against our evaluation framework, we can quantify exactly how much each change helps.

## What you'll learn

Through hands-on experimentation with a personal assistant chatbot, you'll discover how to:


1. Leverage System Prompts
- Write effective prompts that explain tool usage patterns
- Help models understand user workflows and preferences
- Validate prompt improvements with metrics

2. Design Few-Shot Examples 
- Create examples that demonstrate correct tool combinations
- Target specific failure modes identified in testing
- Balance example diversity and relevance


3. Measure Improvements
- Compare performance before and after changes
- Track precision and recall across different approaches
- Make data-driven decisions about prompting strategies

By the end of this notebook, you'll understand how to systematically improve tool selection accuracy through better prompting, and how to measure the impact of these changes using objective metrics.

## System Prompts

By adding a system prompt for users to outline their specific workflow and tool usage, our model can handle a greater variety of users and their specific tool usage patterns.

Let's see this in action below where we add the user provided system prompt to our prompt template.

In [3]:
import instructor
from helpers import load_commands, load_queries, Command, SelectedCommands

user_system_prompt = """
I work as a software engineer at a company. When it comes to work, we normally track all outstanding tasks in Jira and handle the code review/discussions in github itself. 

refer to jira to get ticket information and updates but github is where code reviews, discussions and any other specific code related updates are tracked. Use the recently updated issues command to get the latest updates over search. 

for todos, i use a single note in apple notes for all my todos unless i say otherwise. Obsidian is where I store diagrams, charts and notes that I've taken down for things that I'm studying up on. Our company uses confluence for documentation, wikis, release reports, meeting notes etc that need to be shared with the rest of the team. Notion I use it for financial planning, tracking expenses and planning for trips. I always use databases in notion.

For messaging apps, I tend to just use discord for chatting with my friends when we game, i use microsoft teams for communicating with colleague about spcifically work related matters and iMessage for personal day to day stuff (Eg. coordinate a party, ask about general things in a personal context)
"""


async def generate_commands_with_system_prompt(
    query: str,
    client: instructor.AsyncInstructor,
    commands: list[Command],
    user_system_prompt: str,
):
    response = await client.chat.completions.create(
        messages=[
            {
                "role": "system",
                "content": """
                You are a helpful assistant that can execute commands in response to a user query. You have access to the following commands:
                
                <commands>
                {% for command in commands %}
                - {{ command.key }} : {{ command.command_description }}
                {% endfor %}
                </commands>

                You must select at least one command to be called.

                Here is some information about how the user uses each extension. Remember to find a chat before sending a message.

                <user_behaviour>
                {{ user_behaviour }}
                </user_behaviour>
                """,
            },
            {
                "role": "user",
                "content": query,
            },
        ],
        model="gpt-4o",
        response_model=SelectedCommands,
        context={"commands": commands, "user_behaviour": user_system_prompt},
    )
    return response.selected_commands

/Users/ivanleo/Documents/coding/systematically-improving-rag/cohort_2/.venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


Let's now run our evaluations again to see how it performs.

In [4]:
import instructor
from openai import AsyncOpenAI

commands = load_commands("raw_commands.json")
queries = load_queries(commands, "queries.jsonl")
client = instructor.from_openai(AsyncOpenAI())
await generate_commands_with_system_prompt(
    "notify #engineering that I'm running late for our meeting and i'll be there in 20 minutes",
    client,
    commands,
    user_system_prompt,
)

[UserCommand(key='microsoft-teams.findChat', arguments=[UserCommandArgument(title='Chat/Channel Name', value='engineering')])]

In [ ]:
from braintrust import Score, Eval
from helpers import calculate_precision, calculate_recall


def evaluate_braintrust(input, output, **kwargs):
    return [
        Score(
            name="precision",
            score=calculate_precision(output, kwargs["expected"]),
        ),
        Score(
            name="recall",
            score=calculate_recall(output, kwargs["expected"]),
        ),
    ]


commands = load_commands("raw_commands.json")
queries = load_queries(commands, "queries.jsonl")


client = instructor.from_openai(AsyncOpenAI())
commands = load_commands("raw_commands.json")
queries = load_queries(commands, "queries.jsonl")


async def task(query, hooks):
    resp = await generate_commands_with_system_prompt(
        query, client, commands, user_system_prompt
    )
    hooks.meta(
        input=query,
        output=resp,
    )
    return [item.key for item in resp]


results = await Eval(
    "function-calling",
    data=[
        {
            "input": row["query"],
            "expected": row["labels"],
        }
        for row in queries
    ],
    task=task,
    scores=[evaluate_braintrust],
)

Experiment week-6-fixes-1741045428 is running at https://www.braintrust.dev/app/567/p/function-calling/experiments/week-6-fixes-1741045428
function-calling (data): 51it [00:00, 87488.55it/s]


function-calling (tasks):   0%|          | 0/51 [00:00<?, ?it/s]

/var/folders/ws/q_m6c6qs3n553603dk_zvrgc0000gn/T/ipykernel_28820/4070492607.py:31: DeprecationWarning: meta() is deprecated. Use the metadata field directly instead.
  hooks.meta(



=========================SUMMARY=========================
week-6-fixes-1741045428 compared to week-6-fixes-1741045394:
51.14% (+09.14%) 'recall'    score	(18 improvements, 11 regressions)
62.08% (+18.14%) 'precision' score	(21 improvements, 9 regressions)

1741045428.53s start
1741045430.89s end
2.35s (-31.09%) 'duration'	(32 improvements, 19 regressions)

See results for week-6-fixes-1741045428 at https://www.braintrust.dev/app/567/p/function-calling/experiments/week-6-fixes-1741045428


By providing a system prompt, we saw a ~44% increase in precision from 0.43 to 0.62 whereas for recall, we saw a ~21% increase from 0.42 to 0.51

This is a significant improvement and shows that providing a system prompt can help the model understand how the user uses each tool. 

Better yet, using system prompts allow our model to be more flexible and handle a greater variety of users that may have different ways of interacting with the tools. 

### Comparing System Prompts with Baseline

Now that we've seen a overall improvement across the board, let's look at what specific queries our model is having issues with. Let's do so by computing the same metrics as we did in our previous notebook

In [11]:
import pandas as pd
from helpers import calculate_per_tool_recall, calculate_precision_recall_for_queries

df = pd.DataFrame(
    [
        {
            "query": row.input,
            "expected": row.expected,
            "actual": row.output,
        }
        for row in results.results
    ]
)

df = calculate_precision_recall_for_queries(df)
df.head(10)

query  \
0  create a grocery list note with Milk, eggs, br...   
1  Let's create a new release post about our late...   
2  find weather taiwan december and generate a sh...   
3  Set my status to 'Working from home today, cat...   
4  any security alerts raised since we upgraded o...   
5        Tell mum i'll be back for dinner around 7pm   
6  just booked the latest accomodations for tokyo...   
7  search messages Gregory modal, need to find th...   
8  add a reminder to my todos to buy some groceri...   
9  pull up munich plans, send mike the airbnb lin...   

                                            expected  \
0  [apple-notes.new, apple-notes.add-text, apple-...   
1  [confluence-search.new-blog, jira.active-sprin...   
2  [google-search.index, apple-notes.new, apple-n...   
3                        [microsoft-teams.setStatus]   
4                      [github.unread-notifications]   
5          [imessage.findChat, imessage.sendMessage]   
6                      [notion.create-database-page]   
7                   [microsoft-teams.searchMessages]   
8                             [apple-notes.add-text]   
9  [notion.search-page, imessage.findChat, imessa...   

                                              actual  precision  recall  \
0            [apple-notes.new, apple-notes.menu-bar]       1.00    0.67   
1  [jira.recently-updated-issues, microsoft-teams...       0.50    0.25   
2           [google-search.index, apple-notes.index]       0.50    0.33   
3  [microsoft-teams.setStatus, microsoft-teams.fi...       0.33    1.00   
4  [jira.recently-updated-issues, github.notifica...       0.00    0.00   
5                                [imessage.findChat]       1.00    0.50   
6                      [notion.create-database-page]       1.00    1.00   
7                         [microsoft-teams.findChat]       0.00    0.00   
8          [apple-notes.index, apple-notes.add-text]       0.50    1.00   
9                               [notion.search-page]       1.00    0.33   

  CORRECT  
0       N  
1       N  
2       N  
3       N  
4       N  
5       N  
6       Y  
7       N  
8       N  
9       N

In [12]:
df_per_tool = calculate_per_tool_recall(df)
df_per_tool.sort_values(by='Recall', ascending=True).head(15)

Tool  Correct Identification  Total Targets  \
29      confluence-search.add-text                       0              1   
26     github.unread-notifications                       0              5   
20            confluence-search.go                       0              1   
19          discord.searchMessages                       0              2   
33  microsoft-teams.searchMessages                       0              1   
32     microsoft-teams.sendMessage                       0             13   
36             discord.sendMessage                       0              3   
30            imessage.sendMessage                       0              5   
17      confluence-search.new-page                       0              1   
28      confluence-search.new-blog                       0              2   
12                jira.open-issues                       0              1   
23              jira.search-issues                       1              3   
3         confluence-search.search                       1              3   
9                  apple-notes.new                       1              2   
8              jira.active-sprints                       1              2   

    Recall  
29    0.00  
26    0.00  
20    0.00  
19    0.00  
33    0.00  
32    0.00  
36    0.00  
30    0.00  
17    0.00  
28    0.00  
12    0.00  
23    0.33  
3     0.33  
9     0.50  
8     0.50

From the table above, we can see that some potential function calls with a high discrepancy between the expected calls and actual calls are 

1. `microsoftTeams.sendMessage` and `imessage.sendMessage`
2. `github.unreadNotifications`
3. `discord.findChat` and `discord.sendMessage`

We'll grab some of these examples where a specific tool call didn't get executed below


In [15]:
def get_mismatched_examples_for_tool(df, tool_substring, num_examples=3):
    """
    Filter dataframe for rows where a specific tool substring appears in expected tools
    and the actual output doesn't match the expected output.
    
    Args:
        df: DataFrame containing the data
        tool_substring: String to search for in the expected tools
        num_examples: Number of examples to display (default: 3)
        
    Returns:
        DataFrame containing filtered examples where actual != expected
    """
    # Filter for rows where the tool substring is in expected tools
    filtered_df = df[df["expected"].apply(lambda x: any(tool_substring in item for item in x))]
    
    # Filter for rows where actual doesn't match expected
    mismatched_df = filtered_df[filtered_df.apply(lambda row: set(row['expected']) != set(row['actual']), axis=1)]
    
    # Return the top examples
    return mismatched_df.head(num_examples)

# Set display width to maximum
pd.set_option('display.max_colwidth', None)
pd.set_option('display.width', None)

# Example usage for unread-notifications
unread_notification_examples = get_mismatched_examples_for_tool(df, "unread-notifications",num_examples=5)
unread_notification_examples

query  \
4   any security alerts raised since we upgraded our nextjs dependencies over to 14.2.0?    
22                                        any more prs or security alerts to worry about?   
48                      check if there are any dependency vulnerabilities raised recently   
49                                  did anyone comment on the pr for the performance fix?   
50                               pull up those security alerts and ping the security team   

                                                                                expected  \
4                                                          [github.unread-notifications]   
22                                                         [github.unread-notifications]   
48                                                         [github.unread-notifications]   
49                                                         [github.unread-notifications]   
50  [github.unread-notifications, microsoft-teams.findChat, microsoft-teams.sendMessage]   

                                                  actual  precision  recall  \
4   [jira.recently-updated-issues, github.notifications]        0.0    0.00   
22       [github.my-pull-requests, github.notifications]        0.0    0.00   
48  [jira.recently-updated-issues, github.notifications]        0.0    0.00   
49                             [github.my-pull-requests]        0.0    0.00   
50      [github.notifications, microsoft-teams.findChat]        0.5    0.33   

   CORRECT  
4        N  
22       N  
48       N  
49       N  
50       N

In [16]:
unread_notification_examples = get_mismatched_examples_for_tool(df, "unread-notifications",num_examples=5)
unread_notification_examples

query  \
4   any security alerts raised since we upgraded our nextjs dependencies over to 14.2.0?    
22                                        any more prs or security alerts to worry about?   
48                      check if there are any dependency vulnerabilities raised recently   
49                                  did anyone comment on the pr for the performance fix?   
50                               pull up those security alerts and ping the security team   

                                                                                expected  \
4                                                          [github.unread-notifications]   
22                                                         [github.unread-notifications]   
48                                                         [github.unread-notifications]   
49                                                         [github.unread-notifications]   
50  [github.unread-notifications, microsoft-teams.findChat, microsoft-teams.sendMessage]   

                                                  actual  precision  recall  \
4   [jira.recently-updated-issues, github.notifications]        0.0    0.00   
22       [github.my-pull-requests, github.notifications]        0.0    0.00   
48  [jira.recently-updated-issues, github.notifications]        0.0    0.00   
49                             [github.my-pull-requests]        0.0    0.00   
50      [github.notifications, microsoft-teams.findChat]        0.5    0.33   

   CORRECT  
4        N  
22       N  
48       N  
49       N  
50       N

In [17]:
send_message_examples = get_mismatched_examples_for_tool(df, "imessage.sendMessage",num_examples=5)
send_message_examples

query  \
5                                 Tell mum i'll be back for dinner around 7pm   
9   pull up munich plans, send mike the airbnb link to the accoms on the 22nd   
27          Message David to ask if he's still up for basketball this weekend   
28                     Send Kevin a message asking if he still has my charger   
29                           Tell Alex I'm running 15 minutes late for brunch   

                                                         expected  \
5                       [imessage.findChat, imessage.sendMessage]   
9   [notion.search-page, imessage.findChat, imessage.sendMessage]   
27                      [imessage.findChat, imessage.sendMessage]   
28                      [imessage.findChat, imessage.sendMessage]   
29                      [imessage.findChat, imessage.sendMessage]   

                  actual  precision  recall CORRECT  
5    [imessage.findChat]        1.0    0.50       N  
9   [notion.search-page]        1.0    0.33       N  
27   [imessage.findChat]        1.0    0.50       N  
28   [imessage.findChat]        1.0    0.50       N  
29   [imessage.findChat]        1.0    0.50       N

In [18]:
send_message_examples = get_mismatched_examples_for_tool(df, "discord.sendMessage",num_examples=5)
send_message_examples

query  \
12                   let's go crack open that new raid, set status to dnd and ping #bois   
44                               tell #team-alpha I'll be late for tonight's dungeon run   
45  drop a message in #guild chat that I'm taking a break and set status to idle for now   

                                                            expected  \
12  [discord.setStatus, discord.searchMessages, discord.sendMessage]   
44                           [discord.findChat, discord.sendMessage]   
45        [discord.setStatus, discord.findChat, discord.sendMessage]   

                 actual  precision  recall CORRECT  
12  [discord.setStatus]        1.0    0.33       N  
44   [discord.findChat]        1.0    0.50       N  
45   [discord.findChat]        1.0    0.33       N

### Few Shot Prompting

Once we've identified potential problem areas - like the model failing to find findChat - few shot examples can explicitly demonstrate these commands used in context. 

For instance, we can show a few examples of how to use the `findChat` command with a `sendMessage` command. A natural fit here could be to grab some content from an internal documentation site like `confluence` and then sending it over to a chat.

```
<query>generate release notes for the tickets closed in our current sprint and send the link over to the #product channel ahead of time so they know what's coming</query>
<commands>
    confluence-search.new-blog,
    confluence-search.add-text,
    microsoft-teams.findChat,
    microsoft-teams.sendMessage
</commands>
```

We could also be inventive and use the `searchMedia` command alongside a normal `searchNoteCommand` to show the model how each command differs.

```
<query>Can you grab my notes and sketches which I put together about cross-attention?</query>
<commands>
    obsidian.searchMedia,
    obsidian.searchNote
</commands>
```

Including these concrete examples in the prompt teaches the model the correct sequence of steps and drastically reduces the chances it calls the wrong command.”

In [19]:
async def generate_commands_with_prompt_and_examples(
    query: str,
    client: instructor.AsyncInstructor,
    commands: list[Command],
    user_behaviour: str,
):
    response = await client.chat.completions.create(
        messages=[
            {
                "role": "system",
                "content": """
You are a helpful assistant that can execute commands in response to a user query. Only choose from the commands listed below. 

You have access to the following commands:

<commands>
{% for command in commands %}
<command>
    <command key>{{ command.key }}</command key>
    <command description>{{ command.command_description }}</command description>
</command>
{% endfor %}
</commands>

Select between 1-4 commands to be called in response to the user query.

Here is some information about how the user uses each extension. Remember to find a chat before sending a message.

<user_behaviour>
{{ user_behaviour }}
</user_behaviour>

Here are some past examples of queries that the user has asked in the past and the keys of the commands that were expected to be called. These provide valuable context and so look at it carefully and understand why each command was called, taking into account the user query below and the user behaviour provided above.

<examples>
    <example>
        <query>Compile any new outstanding PRs that have been reviewed recently with any vulnerabilities that have been reported and send them in a message to the engineering team in the #engineering channel to fix it before our next release</query>
        <commands>
            github.unread-notifications
            microsoft-teams.findC hat
            microsoft-teams.sendMessage
        </commands>
    </example>
    <example>
        <query>Can you text Philip a link to the notion document for our trip to Taiwan next week?</query>
        <commands>
            notion.search
            imessage.findChat
            imessage.sendMessage
        </commands>
    </example>
    <example>
        <query>Can you pull up the visualisation I made to show how our D&D dungeon map layout works and then forward it to the party members in #gaming?</query>
        <commands>
            obsidian.searchMedia
            discord.findChat
            discord.sendMessage
        </commands>
    </example>
</examples>
                """,
            },
            {
                "role": "user",
                "content": query,
            },
        ],
        model="gpt-4o",
        response_model=SelectedCommands,
        context={"commands": commands, "user_behaviour": user_behaviour},
    )
    return response.selected_commands


In [20]:
from braintrust import Score, Eval
from helpers import calculate_precision, calculate_recall



def evaluate_braintrust(input, output, **kwargs):
    return [
        Score(
            name="precision",
            score=calculate_precision(output, kwargs["expected"]),
        ),
        Score(
            name="recall",
            score=calculate_recall(output, kwargs["expected"]),
        ),
    ]


commands = load_commands("raw_commands.json")
queries = load_queries(commands, "queries.jsonl")


client = instructor.from_openai(AsyncOpenAI())
commands = load_commands("raw_commands.json")
queries = load_queries(commands, "queries.jsonl")


async def task(query, hooks):
    resp = await generate_commands_with_prompt_and_examples(
        query, client, commands, user_system_prompt
    )
    hooks.meta(
        input=query,
        output=resp,
    )
    return [item.key for item in resp]


results = await Eval(
    "function-calling",
    data=[
        {
            "input": row["query"],
            "expected": row["labels"],
        }
        for row in queries
    ],
    task=task,
    scores=[evaluate_braintrust],
)

Experiment week-6-fixes-1741045780 is running at https://www.braintrust.dev/app/567/p/function-calling/experiments/week-6-fixes-1741045780
function-calling (data): 51it [00:00, 39199.10it/s]


function-calling (tasks):   0%|          | 0/51 [00:00<?, ?it/s]

/var/folders/ws/q_m6c6qs3n553603dk_zvrgc0000gn/T/ipykernel_28820/4172904249.py:32: DeprecationWarning: meta() is deprecated. Use the metadata field directly instead.
  hooks.meta(



=========================SUMMARY=========================
week-6-fixes-1741045780 compared to week-6-fixes-1741045428:
80.59% (+29.45%) 'recall'    score	(22 improvements, 1 regressions)
77.76% (+15.69%) 'precision' score	(16 improvements, 3 regressions)

1741045780.58s start
1741045783.62s end
3.03s (+67.38%) 'duration'	(9 improvements, 42 regressions)

See results for week-6-fixes-1741045780 at https://www.braintrust.dev/app/567/p/function-calling/experiments/week-6-fixes-1741045780


## Conclusion

In this notebook, we've shown how simple techniques like few-shot prompting and system prompts can significantly boost model performance in tool selection. By providing a system prompt, we saw a ~44% increase in precision from 0.43 to 0.62 whereas for recall, we saw a ~21% increase from 0.42 to 0.51. We were then able to use few-shot prompts to push the precision to 0.78 and the recall to 0.81, representing a ~81% increase in precision and ~93% increase in recall from our baseline of 0.43 and 0.42 respectively.

These gains came from two key insights : clear system prompts help models understand tool usage patterns (like using Teams for work vs Discord for gaming), while targeted examples prevent common mistakes like forgetting to find a chat before sending a message.

This reflects the systematic pattern we've followed throughout the course. Each week started by defining clear metrics to optimize - whether that was MRR and recall for retrieval (Week 1), recall and MRR for metadata filtering in Week 5, or precision metrics for tool selection here in Week 6. With these metrics in place, we could use synthetic data to rapidly test improvements and validate the improvements that our changes have on the system.

As we deploy these systems to production, this data-driven approach becomes even more crucial. By collecting real user feedback through UI elements like thumbs up/down ratings and establishing clear evaluation metrics early, teams can quantify the impact of each change they make. This systematic strategy - defining metrics, using synthetic data for rapid testing, then validating with real user feedback - provides a reliable framework for improving RAG systems even as they grow more complex with multiple data sources, tools and retrieval methods.